Fix blockstate id map calculation.

gigaherz opened the pull request:
Not sure if it's the best approach or fully correct, but it seems to work.

gigaherz committed:
Fix blockstate id map calculation.

jamierocks commented on src/main/java/net/minecraftforge/registries/GameData.java:
```diff
@@ -52,12 +52,7 @@
-import java.util.List;
-import java.util.Locale;
-import java.util.Map;
+import java.util.*;
```
please no

liach commented on patches/minecraft/net/minecraft/block/Block.java.patch:
```diff
@@ -181,7 +181,15 @@
+       func_196254_a("structure_block", new BlockStructure(Block.Builder.func_200949_a(Material.field_151573_f, MapColor.field_197656_x).func_200948_a(-1.0F, 3600000.0F)));
+       field_149771_c.func_177776_a();
+ 
++      if(false)
```
You should add a comment about where this code went.

liach commented:
1. How is the old calculation wrong?
2. Is there anything wrong with skipping unused ids left blank by registry substitution?

gigaherz commented:
We discussed some concerns about this implementation, over on discord. We need to make sure that vanilla clients still work, even after substituting vanilla blocks, and that modded clients have their IDs matching the server.

Consider it WIP in its current state.

gigaherz commented:
Okay now that it works:

>1. How is the old calculation wrong?

The block IDs are generally consecutive (not necessarily, but generally). This means if you have block 35 with 2 states, they would have got IDs 35+36. Then comes block 36, and also takes ID 36, then 37,38, ...

It would have only worked if all the blocks had exactly one valid state.

>2. Is there anything wrong with skipping unused ids left blank by registry substitution?

Not necessarily, but in the spirit of avoiding desyncs when dealing with vanilla clients, modded clients on vanilla servers, and generally in any other multiplayer situation, we decided to go with post-processing the registries instead of calculating IDs on the fly, so that concern is irrelevant now.

The only thing I'm not convinced of, is this exception throwing here: https://github.com/MinecraftForge/MinecraftForge/pull/5279/files#diff-817b222bf5667eecb5a063fb7224b8c6R297

I don't know if we should warn instead of crash with invalid vanilla replacements, if we should warn about modded replacements on top of crashing with vanilla, or if we should just crash always.

bs2609 commented on src/main/java/net/minecraftforge/registries/GameData.java:
```diff
@@ -277,62 +277,34 @@ void clear()
-
             if (oldBlock != null)
             {
-                @SuppressWarnings("unchecked")
```
This block of code shouldn't be removed, it's nothing to do with the old state<->meta logic.

gigaherz replied:
Note that all of this was commented out. I just removed the commented out code, but git matched it because it happens to have the same condition.

bs2609 replied:
I am aware that it was commented out, but as it is not replaced or superseded, I think that that may have been a simple mistake. In any case, it needs clarifying either way.

gigaherz replied:
Undid this removal. You are right that this was unrelated.

gigaherz committed:
Improve replacement test to match blockstate string representations, not just the number of valid states.
Remove pointless BitSet usage, and restore commented-out code that isn't related to this PR.

gigaherz committed:
Fix logic mistake.

bs2609 commented:
Missing a call to the callback from `revertToFrozen`?

gigaherz committed:
Missing bake.

gigaherz commented:
Yup. Added.

bs2609 commented:
It's worth noting that `ClearableObjectIntIdentityMap.clear` now also needs to reset `nextId` to 0.

gigaherz commented:
Err... yes. Worst thing is I did come across that when I was doing the initial implementation, but it left my mind yesterday.

gigaherz committed:
Reset nextId when clearing the id map.

bs2609 commented on patches/minecraft/net/minecraft/block/Block.java.patch:
```diff
@@ -181,7 +181,15 @@
+       func_196254_a("structure_block", new BlockStructure(Block.Builder.func_200949_a(Material.field_151573_f, MapColor.field_197656_x).func_200948_a(-1.0F, 3600000.0F)));
+       field_149771_c.func_177776_a();
+ 
++      if(false) // Processed in GameData.BlockCallbacks#onBake
```
The previous patch here deleted this block rather than just skipping over it, if that was an intentional decision then this should probably do the same.

LexManos replied:
Ya there are times where we use the patches as documentation of deleted/moved sections.

gigaherz committed:
Convert registry construction to use the data directly from builder instead of the invididual parameters.

gigaherz committed:
Move AT changes to the new file and remove old file that isn't used anymore.

gigaherz committed:
Fix missing license headers.

gigaherz committed:
Improve logging.
